### PR TITLE
feature/S4-75: Add API authentication on endpoints used by external APIs/services.

### PR DESCRIPTION
**Pull request by matthewdoreilly**
- Add API authentication on endpoints used by external APIs/services.
- Add pull request template.

Relates to https://github.com/companieshouse/taf-api-karate/pull/190.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/config/SecurityConfig.java [hunk 1]
```diff
@@ -4,30 +4,39 @@
...
 import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+import uk.gov.companieshouse.api.interceptor.InternalUserInterceptor;
 import uk.gov.companieshouse.api.interceptor.TokenPermissionsInterceptor;
 import uk.gov.companieshouse.interceptor.DissolutionTokenPermissionsInterceptor;
 
 @Configuration
 public class SecurityConfig implements WebMvcConfigurer {
 
-    private static final String URI_PATTERN = "/dissolution-request/*";
+    private static final String URI_PATTERN = "/dissolution-request/**";
```
Comment (matthewdoreilly) [inline src/main/java/uk/gov/companieshouse/config/SecurityConfig.java:14]: `/dissolution-request/*` will check endpoints within one path after `/dissolution-request/` (i.e., before the next "/"), such as `/dissolution-request/{company-number}`, but not `/dissolution-request/{company-number}/payment`.
So this should be `/dissolution-request/**` so anything after `/dissolution-request/` is checked.

File: src/test/java/uk/gov/companieshouse/controller/PaymentControllerTest.java [hunk 2]
```diff
@@ -37,6 +38,14 @@
     private static final String PAYMENT_URI = "/dissolution-request/{company-number}/payment";
```
Comment (peadarkelly) [inline src/test/java/uk/gov/companieshouse/controller/PaymentControllerTest.java:38]: could we add some coverage to `POST /dissolution-request/submit` now that it is in master?

File: src/main/java/uk/gov/companieshouse/config/SecurityConfig.java [hunk 3]
```diff
@@ -4,30 +4,39 @@
...
+    private static final String[] TOKEN_PERMISSION_AUTH_EXCLUDE_LIST = {
             "/dissolution-request/healthcheck",
-            "/dissolution-request/{company-number}/payment",
+            "/dissolution-request/{company-number:[0-9]+}/payment",
             "/dissolution-request/submit"
     };
 
+    private static final String[] API_KEY_PERMISSION_AUTH_EXCLUDE_LIST = {
+            "/dissolution-request/healthcheck",
+            "/dissolution-request/{company-number:[0-9]+}"
```
Comment (matthewdoreilly) [inline src/main/java/uk/gov/companieshouse/config/SecurityConfig.java:24]: Previously, `/dissolution-request/{company-number}` was actually excluding `/dissolution-request/submit` from the API Key Interceptor as well (they were both being treated as the same URI pattern).
So needed to update our `company-number` path variable to say this data type will be a number.
More info on this here, https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/util/pattern/PathPattern.html

Comment (peadarkelly) [reply src/main/java/uk/gov/companieshouse/config/SecurityConfig.java:24]: This isn't true, LLDS01 company numbers start with LL

**Review by peadarkelly — APPROVED** [review]

**Pull request merged**